### ISSUE TITLE
On-demand Rideshare Service Example App for firebase cloud Functions (Uber/Lyft Clone)

### ISSUE THREAD

**Issue by Radecom** [firebase/functions-samples#603]
I am designing a clone uber system with Firestore, Cloud Messaging and Functions of Firebase. I have a function that runs when a new order is created (OrdersRef.onCreate ()).

[Diagram Rideshare Dispatch](https://user-images.githubusercontent.com/45552664/63304132-a3815580-c2a7-11e9-9e92-0f0259df2005.jpg)

Obviously, I execute the tasks of:

Get the currently available cars.
For each available car, I get the distance between the location of the car and the pick-up point of the order.
I order the cars by the distance of their location and the point of collection of the order.
*** I have the exception if the array of available cars is empty, the function ends.*

So far everything is going well. But here you ask me, what is the best way to carry out the car dispatch?

I can think of 2 ways:

The first would be to send a notification to the nearest driver with the order data, wait 9 sec (sleeping the cloud function with a sleep function), after waiting 9 seconds, make a new query to get the current status of the order, and if it has not yet been taken, I send a new notification to the next driver ... and so on until I run out of my array of available cars. ** We have to remember that in cloud functions (backend) queries to Firestore are promises and not observable as in frontend frames (angular).

The second would be to schedule notifications and queries with pub/sub and make a review similarly ... Send a notification to the nearest driver with the order data, and program with Pub / Sub a function that will check in the next 9 seconds if the order has already been taken, if I do not send the notification to the next driver and reschedule a new function with Pub / Sub to recheck ... and so on until my array of available cars runs out.

The problem with the first form is that the function can last a long time in execution.

The problem with the second form is that a function can be converted into many subfunctions.

My question here is: What is the best way to do this?

**Could you give an example to guide us like this** [Pubnun](https://www.pubnub.com/blog/build-android-on-demand-rideshare-app-uber-lyft/)**?**

**Comment by IngAjVillalon** [comment]
Such an example would be great.
Please do it!

**Comment by Crewnie** [comment]
What an interesting example!
I join the waiting list

**Comment by FrutasLuna** [comment]
Wow, that would be great,
Yes please and in more detail explain each element of google cloud we would have more idea of ​​how to start.
Because sometimes that is the big problem, we don't even know how to start, and having a guide would be much easier.

**Comment by v1b3m** [comment]
Cloud pubsub sounds like the better option here. Though if you want a
simpler system, the first way is the way to go. Usually, you should go with
the simpler way of implementation, then when your number of users get too
much, that's when you implement a more advanced method. I've implemented
something similar and I use arrays, plus computers are really fast, so it
might not really take as long as you think it will, plus it's ok for
clients to wait for some time.
Hope this helps.

**Comment by desarrollotrimax** [comment]
Yes please, it would help us a lot, especially to have an overview of how to combine all the cloud functions of google cloud.

**Comment by olaide-ekolere** [comment]
@Radecom I also faced this issue and these were the steps I used in resolving it.
1. Available driver coordinates are stored always stored in a collection and updated by the driver app.
2. When a rider makes a ride request is made, in the onCreate, only Y number of available drivers within X km radius of the pick-up point based on distance from the pickup point will be sent push notifications. if no drivers are available, a notification is sent to the rider to try again in a few minutes.
3. The first available driver to accept the request gets it and a notification is sent to the rider. The accept ride function does not allow drivers to accept a ride request after Z minutes from request time.
4. The rider app quits the request after Z minutes if no driver has accepted it.
